src/Base/UniqueNameManager.h

```diff
@@ -0,0 +1,157 @@
+ *   Lesser General Public License for more details.                       *
+ *                                                                         *
+ *   You should have received a copy of the GNU Lesser General Public      *
+ *   License along with FreeCAD. If not, see                               *
+ *   <https://www.gnu.org/licenses/>.                                      *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#ifndef SRC_BASE_UNIQUENAMEMANAGER_H_
+#define SRC_BASE_UNIQUENAMEMANAGER_H_
+
+#ifndef FC_GLOBAL_H
+#include <FCGlobal.h>
+#endif
+#include <vector>
+#include <string>
+#include <set>
+#include <map>
+#include <algorithm>
+#include <utility>
+#include <tuple>
+
+// ----------------------------------------------------------------------------
+
+namespace Base
+{
+/// @brief
+/// This class acts as a multiset of strings with the capability of generating
+///     other names that are not in the collection. Unique names are generated
+///     by inserting digits into a base name. Normally these are appended to the
+///     base name, but a derived class can override getNameSuffixStartPosition
+///     to place the digits elsewhere in the base name.
+// This class does not support copy-/move-construction
+// NOLINTNEXTLINE(cppcoreguidelines-special-member-functions)
+class BaseExport UniqueNameManager
+{
+protected:
+    // This method returns a reverse const iterator on name that references the position
+    // of the start of the suffix (or name.crbegin() if no suffix).
+    // It must return the same suffix length (name.crbegin() - returnValue) for both
+    // unique names (one containing digits) and the corresponding base name (with no digits).
+    virtual std::string::const_reverse_iterator
+    getNameSuffixStartPosition(const std::string& name) const
+    {
+        return name.crbegin();
+    }
+
+private:
+    class PiecewiseSparseIntegerSet
+    {
+    public:
+        PiecewiseSparseIntegerSet() = default;
+
+    private:
+        // Each pair being <lowest, count> represents the span of integers from lowest to
+        // (lowest+count-1) inclusive
+        using spanType = std::pair<unsigned int, unsigned int>;
+        // This span comparer class is analogous to std::less and treats overlapping spans as being
+        // neither greater nor less than each other
+        class comparer
+        {
+        public:
+            bool operator()(spanType lhs, spanType rhs) const
+            {
+                // The equality case here is when lhs is below and directly adjacent to rhs.
+                return lhs.second + lhs.first <= rhs.first;
+            }
+        };
+        // Spans is the set of spans. Adjacent spans are coalesced so there are always gaps between
+        // the entries.
+        std::set<spanType, comparer> Spans;
```

Comment:
Member variables usually has first letter lowercase
```suggestion
        std::set<spanType, comparer> spans;
```